Small enhancements to encoder.c

simo5 opened the pull request:
- Streamlines rsa public key printing
- Allow to print public key data from private keys
- Add EC public key printing.
- Match printing style from OpenSSL's existing text encoders

dengert commented on src/encoder.c:
```diff
@@ -665,3 +664,120 @@ const OSSL_DISPATCH p11prov_ec_encoder_spki_der_functions[] = {
+    }
+
+    keysize = p11prov_obj_get_key_size(key);
+    if (class == CKO_PRIVATE_KEY) {
+        BIO_printf(out, "PKCS11 EC Private Key (%lu bits)\n", keysize * 8);
+    } else {
+        BIO_printf(out, "PKCS11 EC Public Key (%lu bits)\n", keysize * 8);
+    }
```
The EC "key size" is based on the field length in bits and thus is the same for the public and private key.  And it does not have to be power of 2, for example: "NIST P-521 elliptic curve, known also as secp521r1 is 521-bit ECC curve" 

What OpenSSL may be referring to is the  number of bytes needed to hold an ecpointQ  which for p-521 is 56 bytes, and 56*8=528 which is not the field size. 

PKCS11 3.0 says:

> Note: CKF_EC_NAMEDCURVE is deprecated with PKCS#11 3.00. It is replaced by CKF_EC_OID. 
In these standards, there are two different varieties of EC defined: 
1. EC using a field with an odd prime number of elements (i.e. the finite field Fp). 
2. EC using a field of characteristic two (i.e. the finite field F2m). 
An EC key in Cryptoki contains information about which variety of EC it is suited for. It is preferable that a  Cryptoki library, which can perform EC mechanisms, be capable of performing operations with the two  varieties of EC, however this is not required. The CK_MECHANISM_INFO structure CKF_EC_F_P flag  identifies a Cryptoki library supporting EC keys over Fp whereas the CKF_EC_F_2M flag identifies a  Cryptoki library supporting EC keys over F2m. A Cryptoki library that can perform EC mechanisms must  set either or both of these flags for each EC mechanism.


OpenSSL has a way to get the field size for example:
https://github.com/OpenSC/OpenSC/blob/master/src/libopensc/pkcs15-prkey.c#L758-L779

dengert replied:
And for OpenSSL 3.x https://www.openssl.org/docs/man3.1/man7/EVP_PKEY-EC.html

simo5 replied:
> This condition is always false as the `class` is set only during initialization and not modified later.

Good catch!

> Do we expect in the future to be able to print some information about private keys (at least the public information + information that it is provided by the pkcs11 so the private information is not available)?

We always fall back to fetch also public info, so we alwways have at least that to print plus the CKA_ID and CKA_LABEL.

In fact these two additional values are the only reason to have these functions at all, if the functions do note exist OpenSSL simply fetches the data on its own and already prints it (without CKA_Id and CKA_LABEL of course).

> Or not at all? Based on that, we should change this as well as the switch above which also allows only the `OSSL_KEYMGMT_SELECT_PUBLIC_KEY` value either to allow other types or convert it to the simple if.

Yes currently we print only the public key, but openssl will happily call this for a private key too if you pass -pubout to the pkey command.
 
> The same for the RSA keys

Ack.

simo5 replied:
@dengert I am confused is the only issue that for some curves (like P-521) it returns slightly incorrect numbers?
Or is there something else?

simo5 replied:
Should be addressed in the latest patch

simo5 commented:
Ok I pushed a commit that changes what we print after I checked more carefully what the openssl encoders do.
See how I changed the tests as well.

simo5 commented:
I added a PEM encoder function to print public keys, this is needed when we prevent exports.
Oddly enough when I coded this I changed p11prov_store_export_object() to export both RSA and EC public keys.
But for some reasons this had the siude effect of making ECDH fail.
Instead always returning error from this function seem to work just fine.
I haven't exactly figured out what goes wrong in OpenSSL in this case, but for now I will just do the simplest thing and neuter this function.

Jakuje commented on src/encoder.c:
```diff
@@ -665,3 +688,115 @@ const OSSL_DISPATCH p11prov_ec_encoder_spki_der_functions[] = {
+        }
+    }
+
+    a = p11prov_obj_get_attr(key, CKA_ID);
+    if (a) {
+        BIO_printf(out, "Key ID:\n");
+        ASN1_buf_print(out, a->pValue, a->ulValueLen, 4);
+    }
+    a = p11prov_obj_get_attr(key, CKA_LABEL);
+    if (a) {
+        BIO_printf(out, "Label: %*s\n", (int)a->ulValueLen, (char *)a->pValue);
+    }
```
just a nit that this part is the same as for the rsa so this might be an opportunity for helper function to avoid code duplication, but I do not insist on this one.

simo5 replied:
Yes, I was almost going to rehaul the whole code, but then scrapped it to keep this PR "small".

Once we get this one merged I may open a new PR with refactoring of the whole thing, as I thought of a way to simplify some of the code when allocations are needed.
